Undeprecate WebAssembly support

white-axe opened the pull request:
This is a pull request to enable Luminol to run in web browsers again by moving most of Luminol's code into a web worker so that Rust synchronization primitives can use `Atomics.wait()` to synchronously block, and handling the filesystem operations and canvas input events in the main thread. The code can be tested in Chromium-based browsers using `trunk serve`, although it is highly recommended to build in release mode as Rust's memory allocator currently runs extremely slowly in debug mode.

Summary of changes made so far to enable WebAssembly support:

- Added a custom egui app runner in src/web/web_worker_runner.rs that can run in a web worker because `eframe::WebRunner` can only run in the main thread.
- Added a web-specific filesystem implementation in src/filesystem/web.rs that uses a loop running on the main thread to asynchronously wait for filesystem commands from the worker thread. All of the filesystem operations except the directory picker are synchronous and blocking on the worker thread only; the main thread only sleeps asynchronously. The directory picker is asynchronous on the worker thread.
- Added Tokio as a dependency in web builds only. We need Tokio channels for communication between the main thread and worker thread because Tokio channel receivers can wait asynchronously for a message to be sent (`std` channels don't have this capability), a requirement for the filesystem loop since we can't block the main thread.
- Refactored the map and tilepicker to rely on `egui::InputState::time` for animation instead of `std::time::Instant` which can only be used in native builds.
- Refactored the tilemap and sprite shaders to optionally use uniforms instead of push constants because push constants are not supported in most web browsers. We should probably also be using uniforms on DX12 as well until wgpu and naga fix their DX12 bugs, but that is outside the scope of this pull request.
- Added const_format as a dependency in all builds, for preprocessing the WGSL files.
- Refactored the tilemap, events and tilepicker to only render onto viewports that are fully contained in the screen. Rendering onto offscreen viewports is explicitly prohibited by the WebGPU standard [as explained in this blog post](https://babylonjs.medium.com/how-to-simulate-out-of-bounds-viewports-when-using-webgpu-or-babylonnative-2280637c0660). Future versions of wgpu will also prevent us from doing this in native builds.
- Added tracing-log as a dependency in web builds only, to catch log messages from egui and wgpu and forward them to tracing, because #43 updated egui to a version that uses log instead of tracing. Native builds already forward log messages to tracing.
- Added indexed_db_futures as a dependency in web builds only, to make it easier to perform the IndexedDB operations needed to store the project directory handles in the recent projects menu.
- Replaced surf with reqwest in all builds because surf in web builds relies on outdated dependencies.
- Added `AudioWrapper` in src/audio/wrapper.rs, which is a wrapper around `Audio` that can be moved to a different thread in case `Audio` isn't `Send`.

Summary of changes still to be made:

- None!

white-axe committed:
WASM test

(cherry picked from commit 11e2cdef65d15d804e67d5e7ce509139be609694)

white-axe committed:
Cleanup the changes from the previous commit

(cherry picked from commit 52016dda50d89cbe531ac2184b2020fc3b8442ba)

white-axe committed:
Set default build target to `wasm32-unknown-unknown`

This is so that rust-analyzer will use this build target when analyzing
code.

(cherry picked from commit ddfa8842a739ed7797f051ae1f01f7602ad4ca01)

white-axe committed:
Enable WASM threading and disable WASM audio

I turned on some compiler flags to enable experimental support for
atomics and shared memory in WebAssembly, because this will probably be
needed if we end up using web workers to handle the filesystem
operations in WebAssembly builds.

Unfortunately, rodio doesn't support web workers, and in fact rodio
crashes Luminol in WebAssembly when I simply turn on those compiler
flags without making any changes to the code!

The audio in WebAssembly builds will have to be reimplemented without
rodio. I don't think this will be too hard, given that all we really use
rodio for is straightforward audio playback...

white-axe committed:
Merge branch 'master' into wasm

white-axe committed:
Disable WASM threading flags

Apparently you can't use shared memory from inside of WebAssembly
without passing extra arguments to wasm-bindgen that Trunk doesn't
support. That's okay, we'll make do without these compiler flags.

We still need coi-serviceworker.js because `Atomics.wait()` can't be
used without it in most browsers.

rodio also still won't work because we're going to run in a web worker,
where the audio APIs that rodio needs aren't available.

white-axe committed:
Revert previous commit

Nevermind, there's no flag incompatibility issue with Trunk.
I'm probably just sleep-deprived...

white-axe committed:
Refactor to not use `eframe::Frame`

I'm making a completely custom egui app runner for WebAssembly builds,
since eframe's web runner makes such widespread assumptions about the
availability of certain JavaScript APIs that are only present outside of
web workers that it's unsalvageable when it comes to getting it to run
in a web worker. (Both egui and wgpu can run in a web worker, though, so
a custom runner should work.)

But I ran into problems because implementing a custom runner requires
creating `eframe::Frame`s which can't be done outside of the eframe
crate.

In the future it would be best to stop using eframe entirely, but I
don't want to also have to rewrite the native runner right now so this
is the best I can do.

white-axe committed:
Merge branch 'master' into wasm

white-axe committed:
Rudimentary web worker rendering

I added a new app runner that can run in a web worker.

The entire program now runs in a web worker, which means we can now use
things like mutexes, channels and other synchronization constructs
without making the browser angry when they block.

If you try it right now, the canvas will be very small and will not
react to user input. These things will be fixed later.

white-axe committed:
Rewrite web worker launcher code in Rust

The WebAssembly code now runs in both the main thread and the worker
thread with the same code and the same memory (i.e. multithreading),
but with different entry points.

The code on the main thread gets some information from the environment
and sends it to the code on the worker thread. The worker thread runs
the actual program.

The purpose of doing this is so that we can write event handlers in
Rust, for things like detecting user input on the canvas from the main
thread and sending it to the worker thread, or sending a request for a
filesystem operation from the worker thread to the main thread.

white-axe committed:
Add a hook to the main thread for detecting screen size changes

The main thread now listens for changes to the screen size and then
sends them to the worker thread so that it can resize the canvas to fill
the screen.

We use a `std::sync::mpsc::Sender` on the main thread to send the new
screen sizes to a `std::sync::mpsc::Receiver` on the worker thread. This
has the benefit of being more convenient than `Worker.postMessage()` and
also is probably faster because we don't have to deal with the
serialization overhead associated with `Worker.postMessage()`.

The Rust senders and receivers internally use WebAssembly's experimental
atomics and shared memory features since I enabled these things in the
compiler flags. They need Cross-Origin Isolation to be enabled to work,
so don't remove the coi-serviceworker.js shim.

white-axe committed:
Web worker runner now responds to mouse events

white-axe committed:
Move event handler code into web_worker_runner.rs

white-axe committed:
Remove duplicate `window` from event handler code

white-axe committed:
Use Tokio channels instead of std channels

We need a channel with an asynchronous `recv()` method in order to send
messages from the worker thread to the main thread, so we might as well
use Tokio channels everywhere.

white-axe committed:
Fix memory leak in `WebWorkerRunner::setup_render_hooks()`

At least I *think* this was a memory leak.

Anyways, the use of `CustomFrame` was somehow causing the rendering to
slow down slightly every frame. So that makes me think there was a
memory leak somewhere.

I couldn't figure out how to fix it properly so I just removed
`CustomFrame` since we don't need it anyways. But the most important
thing is that it runs at full speed now!

white-axe committed:
Add a channel for filesystem operations

This commit adds a channel from the worker thread to the main thread so
that the worker thread can send the main thread requests for filesystem
operations. The results are sent back using a Tokio oneshot channel.

white-axe committed:
Channel senders are now stored as clones instead of by reference

This frees us from having to deal with manually specifying lifetimes
while still otherwise behaving exactly the same.

white-axe committed:
Set `max_texture_side` properly in `egui::RawInput`

white-axe committed:
Implement `FileSystem` for WebAssembly builds

The `File` implementation is not done yet, but it will be soon!

white-axe committed:
`FileSystemCommand` variants are now private

white-axe committed:
Implement `File` for WebAssembly builds

We now have a complete filesystem implementation for web browsers.
Finally.

white-axe committed:
Implement project loading for WebAssembly builds

Project loading is quite slow!

I will be experimenting with a path cache for the filesystem later to
see if that helps with speed.

white-axe committed:
Use egui context's time instead of system time

`std::time::Instant` can't be used in WASM.

white-axe committed:
Also stop using system time in tilepicker.rs

white-axe committed:
Split sprite.wgsl and tilemap.wgsl into two parts

white-axe committed:
Enable wasm-bindgen support for reference types and weakrefs

white-axe committed:
Use uniforms instead of push constants in web builds

white-axe committed:
Remove `USING_PUSH_CONSTANTS` from shader headers

white-axe committed:
Tilemap now changes the projection matrix instead of the viewport rect

Before, the tilemap implemented panning and scaling by simply panning
and scaling the viewport rectangle that it rendered onto. This meant
if any part of the tilemap were offscreen, that part of the viewport
would be offscreen, too.

The WebGPU standard actually specifies that you're not allowed to have
any part of the viewport outside of the screen! (i.e. the intersection
of the viewport rectangle and the screen rectangle must be equal to the
viewport rectangle)

Consequently, most web browsers will not allow you to do this.

The version of wgpu that we currently use, version 0.16.3, allows you to
do this on native builds, but only because the developers forgot to
implement a check for this. It's actually undefined behaviour - some
drivers will allow this and some won't. It will be explicitly disallowed
in a future release of wgpu and already is in their trunk branch.

So, I made a minor change to the tilemap so that its viewport rectangle
always stays on screen. Now, the viewport rectangle is always the same
as the UI rectangle that egui allocates for the tilemap, and the
viewport projection matrix changes instead to handle panning and
scaling.

This should be the last change needed for the map editor to work in
WebAssembly, although it also needs to be applied to the event graphics
and the tilepicker, which is not done yet.

white-axe committed:
Tilemap event graphics now use projection matrix

The event graphics on the tilemap now render using the area allocated by
egui for the map editor as the viewport and handle scaling and panning
by modifying the projection matrix.

The event graphics in the hover tooltips do not do this, they just
render using the tooltip as the viewport because the tooltip doesn't
normally go offscreen. For some reason, egui sometimes draws the tooltip
offscreen for one frame before snapping it back onscreen, so a check is
added to disable rendering the tooltip graphic unless the graphic is
fully onscreen.

white-axe committed:
Tilepicker now uses projection matrix

The map editor should now be working in web builds!
Except you won't be able to use the keyboard because I haven't put the
keyboard handling code into the app runner yet.

white-axe committed:
Move JavaScript bindings into bindings.js and bindings.rs

white-axe committed:
Upload bindings.js and bindings.rs

white-axe committed:
Fix native builds

Native builds can now be built with `cargo build` in addition to web
builds being able to be built with `trunk build`.

I really don't like these hooks, but we need to enable build-std only in
web builds, and Trunk provides us no other way to do this.

white-axe committed:
Apply clippy recommendation

white-axe committed:
Add web event handler for keyboard presses/releases

white-axe committed:
Add web event handler for modifier keys

white-axe committed:
Remove unnecessary field from tile shader.rs

white-axe committed:
Cleanup web keyboard handler code

white-axe committed:
Add web event handler for text input

This is used by egui when typing in text boxes.

white-axe committed:
Add web event handler for scrolling

white-axe committed:
Fix log messages not appearing in web builds

white-axe committed:
Fix handling of `pixels_per_point` in web builds

white-axe committed:
Add web event handler for egui's cursor change requests

white-axe committed:
Refactor web initialization code

This refactors the code for initializing Luminol in web builds to not
use global state that is shared between different parts of the program,
and also moves the `USE_PUSH_CONSTANTS` constant into src/components,
the only place it's used from.

white-axe committed:
Add web event handlers for clipboard and opening URLs

white-axe committed:
Add `Clipboard` feature to web-sys

white-axe committed:
Remove unnecessary `mut` in src/filesystem/web.rs

white-axe committed:
Add handler for RTP loading in web builds

Since we have no way to load from an arbitrary directory without the
user's permission, they're currently loaded from the "RTP" subdirectory
in the project directory. This can be changed in the future if required.

white-axe committed:
Remove another unnecessary `mut` in src/filesystem/web.rs

white-axe committed:
Refactor archiver to work in web builds

I don't have any archives to test with, but I can't think of any reason
why these simple changes wouldn't work. Feel free to prove me wrong!

white-axe committed:
Remove unnecessary `.clone()` from src/main.rs

white-axe committed:
Implement eframe storage and egui memory in web builds

This allows most state to be persisted. Unfortunately, we'll need a
different kind of storage to persist the project directory handles.

white-axe committed:
Implement project directory persistence for web builds

Project directory handles are now stored in IndexedDB so that they can
be recovered after the user closes or refreshes the page.

white-axe committed:
Unregister render loop and event listeners on panic

This prevents the console from being spammed with WebGPU errors in the
event of a panic.

melody-rs commented:
Have you tried using kira for audio instead of rodio?

white-axe commented:
I forgot to update this yesterday but it turns out rodio may be usable, the crash I experienced earlier was exclusively due to a cpal bug that will be fixed by

* https://github.com/RustAudio/cpal/pull/774

so I just patched cpal in Cargo.toml to use the version of cpal from that pull request and it stopped crashing. Still have to run rodio in the main thread, though, so I'll have to make some minor modifications to the audio code.

white-axe committed:
Re-add rodio as a dependency in web builds

white-axe committed:
Save on mouseleave in web builds

white-axe committed:
Restore the fullscreen toggle in native builds

white-axe committed:
Restore playtest and console in native builds

white-axe committed:
Restore the quit button in native builds

white-axe committed:
Implement project creation for web builds

I had to replace surf with reqwest because surf won't compile with
wasm-bindgen > 0.2.84 and eframe won't compile with
wasm-bindgen < 0.2.86.

white-axe committed:
Drop directory handles from IndexedDB on error

white-axe committed:
Implement audio in web builds

I replaced the Symphonia decoders with rodio's default ones because
rodio can't decode OGG files properly when using Symphonia's decoder.

white-axe committed:
Don't include `AudioWrapper` in native builds

white-axe committed:
Use `.removeEntry` to remove files/dirs instead of `.remove`

`FileSystemDirectoryHandle.removeEntry` is the standard way to remove
files and directories. `FileSystemHandle.remove` is nonstandard.

white-axe committed:
Fix `Drop` implementation in `AudioWrapper`

It's possible for the `AudioWrapper` to be dropped on a different thread
than it was created on. This commit handles that case.

melody-rs commented on src/main.rs:
```diff
@@ -118,6 +124,128 @@ fn main() {
+
+    filesystem::web::FileSystem::setup_filesystem_sender(filesystem_tx);
+
+    let web_options = eframe::WebOptions::default();
```
Changing the web options to only include webgl as a backend still makes wgpu use WebGPU?

white-axe replied:
I had to make a custom app runner for web builds and I don't think I included WebGL support. I don't think Luminol would even run in WebGL right now because it has a texture size limit of 2048 instead of 8192. But I could still put WebGL support into the app runner if you want.

melody-rs replied:
Ah, alright! You don't need to. It means I'll have to test this on windows (chrome's linux webgpu support is half-baked) but that's ok!

melody-rs commented on src/main.rs:
```diff
@@ -118,6 +124,128 @@ fn main() {
+        luminol::web::web_worker_runner::panic_hook();
+
+        let report = panic.panic_report(info);
+        web_sys::console::log_1(&js_sys::JsString::from(report.to_string()));
```
We should probably present a popup to the user as well as logging to the console

melody-rs commented on src/filesystem/archiver.rs:
```diff
@@ -21,10 +21,14 @@ use itertools::Itertools;
-pub struct FileSystem {
+pub struct FileSystem<T>
+where
+    T: FileSystemTrait,
```
This should probably be restricted to a file rather than a filesystem (although this was an issue with the original archiver)

melody-rs commented on src/components/mod.rs:
```diff
@@ -22,6 +22,17 @@
+/// Whether or not to use push constants when rendering the map editor. Disabling this will switch
+/// to fallback rendering using uniforms, which is slightly slower but is required for Luminol to
+/// work in web browsers until push constants are standardized in WebGPU.
+pub const USE_PUSH_CONSTANTS: bool = false;
```
Rather than a using a platform defined constant detecting if push constants are supported at runtime (using `wgpu::Adapter::features`) would allow luminol to work on platforms that don't support push constants

melody-rs commented on src/windows/new_project.rs:
```diff
@@ -239,21 +239,17 @@ impl Window {
-            surf::get(url)
-            .middleware(surf::middleware::Redirect::new(10))))
-        .await;
+        let zips = futures::future::join_all(zip_url.iter().map(|url| reqwest::get(*url))).await;
```
Does this work with CORS?

white-axe replied:
Probably, coi-serviceworker.js is set up to use COEP in "credentialless" mode so it should be able to download files from other websites

white-axe committed:
Show an alert box in web builds on panic

white-axe committed:
Fix command names in hooks on Windows

white-axe committed:
Use `wgpu::Adapter::features` to detect push constants support

white-axe committed:
Don't show panic message if COI isn't enabled

white-axe committed:
Archiver filesystem now wraps a file instead of a filesystem

white-axe committed:
Apply clippy recommendations

melody-rs reviewed (commented):
I'm amazed that you got opening recent projects to work!

melody-rs commented on src/filesystem/web.rs:
```diff
@@ -0,0 +1,1082 @@
+impl FileSystemTrait for FileSystem {
+    type File<'fs> = File where Self: 'fs;
+
+    fn open_file(
```
Saving a project has the error "permission denied"
![image](https://github.com/Astrabit-ST/Luminol/assets/54482069/3e9c7079-652d-435c-9d9f-b4b16803f2f2)

melody-rs commented on src/filesystem/project.rs:
```diff
@@ -260,6 +290,41 @@ impl FileSystem {
+                }
+                seen_rtps.push(rtp);
+
+                let path = camino::Utf8PathBuf::from("RTP").join(rtp);
```
RTP loading fails if I have the RTP folder in the project:
![image](https://github.com/Astrabit-ST/Luminol/assets/54482069/3eb555b7-3c91-456b-9e57-31d6095ddf10)
![image](https://github.com/Astrabit-ST/Luminol/assets/54482069/7051846a-eb4b-4e65-82f6-d8d58af409c7)

white-axe committed:
Re-enable sound effect picker in the items window

white-axe committed:
Make some minor stylistic changes to names in src/filesystem/web.rs

white-axe committed:
Fix RTPs not being detected properly in web builds

That's seriously embarassing, haha.

white-axe committed:
Auto-enable `Write` when `Truncate` or `Create` are set in web builds

This probably shouldn't be necessary given that Rust's documentation
states that you need to manually enable `Write` to use `Truncate` or
`Create`, but I guess if it worked before in native then it should be
supported in web builds, too.

white-axe committed:
Web filesystem `remove_dir` is now recursive

This is for consistency with the host filesystem's implementation.

white-axe committed:
Make `project_path` in web builds consistent with native

white-axe committed:
Optimize path cache to avoid double-loading the web RTP folder

white-axe committed:
Move `skip` variable from previous commit outside of the loop

white-axe committed:
Fix crash when loading from nonexistent folder in web builds

white-axe commented:
Does project saving even work properly on native yet? Every time I try to save it just says "os error 22".

melody-rs commented:
Huh, apparently not! That's super strange.
